### DISCUSSION TITLE
Automatic transformation of quantum circuits into Pulser code

### DISCUSSION THREAD

**Discussion by LaurentAjdnik** [pasqal-io/Pulser#230] [Ideas] [answered]
Probably a crazy, or even meaningless, idea but...

In https://arxiv.org/abs/2104.15044, the guided walkthrough (§3.2) shows how to turn a typical quantum circuit into Pulser code.

Some gates can be "Pulsered" in a straightforward way. Some others (like CNOT) need to be transformed first, in a form compatible with neutral-atom devices.

Would it be possible and interesting to add a feature that inputs **any circuit encoded with the most common tools** ([Cirq](https://quantumai.google/cirq), [Qiskit](https://qiskit.org/), [QuTip](https://qutip.org/)...) and **transform it automatically into a Registrer/Sequence/Pulses set** ?

The result would certainly depend on the chosen device specifications. But we could imagine an intermediate representation that would be (as) device-agnostic (as possible), and can later be applied to a given device.

It would be some kind of superlayer to the existing features.

Note: Some tools already manipulate circuits created by others. For instance, the brilliant [Mitiq](https://mitiq.readthedocs.io/en/stable/) has support for Braket, Cirq, PyQuil and Qiskit.

**Comment by HGSilveri** [comment, marked as answer]
It's a great idea, and one that we have considered already at Pasqal. Actually, you can already find a frontend for Pasqal devices inside Cirq. We have been working on an intermediate representation that would take the quantum circuit from the frontend all the way to a Pulser sequence for some time now, but this work remains internal at this point. 

It's worth mentioning that it is not just a naive translator of gates to pulses (although it can be that, too), but rather a series of routines that transform the given quantum circuit in a way that tries to leverage the neutral atom devices' architecture as much as possible.

So, to answer your question, it would indeed be possible and interesting to have such a feature, so much so that we are already developing it 😁 

Thanks for the suggestion, it's good to know we are in sync!

**Reply by LaurentAjdnik** [reply]
Hi @HGSilveri, you can mark this as "Answered".
